Data_Engine/Query/LeafNodes.cs

```diff
@@ -0,0 +1,51 @@
+﻿/*
+ * This file is part of the Buildings and Habitats object Model (BHoM)
+ * Copyright (c) 2015 - 2020, the respective contributors. All rights reserved.
+ *
+ * Each contributor holds copyright over their respective contributions.
+ * The project versioning (Git) records all such contribution source information.
+ *                                           
+ *                                                                              
+ * The BHoM is free software: you can redistribute it and/or modify         
+ * it under the terms of the GNU Lesser General Public License as published by  
+ * the Free Software Foundation, either version 3.0 of the License, or          
+ * (at your option) any later version.                                          
+ *                                                                              
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using System.Collections.Generic;
+using BH.oM.Data.Collections;
+using System.Linq;
+using System.ComponentModel;
+using BH.oM.Reflection.Attributes;
+
+namespace BH.Engine.Data
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /****           Public Methods                  ****/
+        /***************************************************/
+        [Description("Extracts a list of leaf nodes from a graph, leaf nodes have no incoming links")]
```

Comment:
That's interesting. In the first example of a tree that came to my mind, leafs would have no **outgoing** links. Take for example a tree representing a population split over multiple levels with the root being the total population. A link could be going from parent to children using the weight as the ratio. Same would go for a water/gas distribution network. 

If I want to be annoying, I could also argue that a leaf node could have both incoming and outgoing links to the same parent.

So it might be safer to return as leaves the nodes that are not connected to any node deeper than themselves.

Reply:
Good points. And useful to see this from another perspective. My use case for this is coming from the decision graph where leaves contribute to the next level up. 

Another classic case I have worked on where its useful to have leaves with only outgoing links is computing flow accumulation in water sheds in river networks. Here the graph can be defined between cells in a landscape linked by negative gradient in topology.  Rainfall from individual cells in a landscape contribute to others

Agreed also that incoming and outgoing links from a leaf are possible. You'll have to try harder than that to annoy 😄. We need to accommodate all these cases. Going to switch to a depth based method of determining a leaf.

Reply:
I think I agree with the first part of @adecler's comment. The most common definition of a `leaf` is a node with no children. If a most depth-wise node has outgoing links to the parent is a just a cyclic graph with no leaves. 

This includes cases where a node is parent of itself as well, [loops](https://en.wikipedia.org/wiki/Loop_(graph_theory)).

I think it would be good to clarify what is `Graph`. To me, the `Graph` class can represent non non-directed cyclic graphs and all its restrictions, either [directed](https://en.wikipedia.org/wiki/Directed_graph) and [acyclic](https://en.wikipedia.org/wiki/Directed_acyclic_graph).